pipeline_dp/dp_computations.py

```diff
@@ -104,3 +205,36 @@ def compute_dp_sum(sum: float, dp_params: MeanVarParams):
     return _add_random_noise(sum, dp_params.eps, dp_params.delta,
                              l0_sensitivity, linf_sensitivity,
                              dp_params.noise_kind)
+
+
+def compute_dp_mean(count: int, sum: float, dp_params: MeanVarParams):
+    """Computes DP mean.
+
+    Args:
+        count: Non-DP count.
+        sum: Non-DP sum.
+        dp_params: The parameters used at computing the noise.
+
+    Raises:
+        ValueError: The noise kind is invalid.
+    """
+    middle = dp_params.middle()
+    normalized_sum = sum - count * middle
+
+    l0_sensitivity = dp_params.l0_sensitivity()
+
+    # Splits the budget equally between the two mechanisms.
+    (sum_eps, sum_delta), (count_eps, count_delta) = equally_split_budget(
+        dp_params.eps, dp_params.delta, 2)
+
+    dp_normalized_sum = _add_random_noise(normalized_sum, sum_eps, sum_delta,
+                                          l0_sensitivity,
+                                          dp_params.linf_sensitivity(
+                                              pipeline_dp.Metrics.MEAN),
+                                          dp_params.noise_kind)
+    dp_count = _add_random_noise(count, count_eps, count_delta, l0_sensitivity,
+                                 dp_params.linf_sensitivity(
+                                     pipeline_dp.Metrics.COUNT),
+                                 dp_params.noise_kind)
+    dp_mean = dp_normalized_sum / dp_count + middle
+    return dp_count, dp_normalized_sum + count * middle, dp_mean
```

Comment:
dp_normalized_sum + count * middle is not differentially private (because count is not differentially private), so we can't return it. As an example, a simple attack would be to calculate dp_normalized_sum = (dp_mean - middle) * dp_count, then we can recover count * middle, and then count directly (since middle is public knowledge).

If you want to release a dp_sum that is different than the one you can get via post-processing (dp_mean * dp_count), then you'll need to spend budget to noise it separately.

Reply:
Thanks for the explanation! It is more clear now. I will modify the function to return dp_mean * dp_count.